renderdoc/driver/vulkan/vk_resources.h

```diff
@@ -1112,6 +1152,524 @@ typedef BitFlagIterator<VkImageAspectFlagBits, VkImageAspectFlags, int32_t> Imag
+  {
+  public:
+    inline SubresourcePairRef &SetState(const ImageSubresourceState &state)
+    {
+      *m_state = state;
+      return *this;
+    }
+  };
+  using ConstSubresourcePairRef = SubresourcePairRefTemplate<const ImageSubresourceState>;
+
+  using SubresourceRangeIter = SubresourceRangeIterTemplate<ImageSubresourceMap, SubresourcePairRef>;
+  using SubresourceRangeConstIter =
+      SubresourceRangeIterTemplate<const ImageSubresourceMap, ConstSubresourcePairRef>;
+
+  inline SubresourceRangeIter RangeBegin(const ImageSubresourceRange &range)
+  {
+    return SubresourceRangeIter(*this, range);
+  }
+  inline SubresourceRangeConstIter RangeBegin(const ImageSubresourceRange &range) const
+  {
+    return SubresourceRangeConstIter(*this, range);
+  }
+  inline SubresourceRangeIter begin() { return RangeBegin(GetImageInfo().FullRange()); }
+  inline SubresourceRangeConstIter begin() const { return RangeBegin(GetImageInfo().FullRange()); }
+  inline SubresourceRangeIter end() { return SubresourceRangeIter(); }
+  inline SubresourceRangeConstIter end() const { return SubresourceRangeConstIter(); }
+  inline size_t size() const { return m_values.size(); }
+};
+
+template <typename Barrier>
+struct BarrierSequence
+{
+  // batches[batchIndex][queueFamilyIndex] = array of barriers to submit to queueFamilyIndex as part
+  // of batchIndex
+  rdcarray<rdcarray<rdcarray<Barrier> > > batches;
+  size_t barrierCount = 0;
+  void Add(uint32_t batchIndex, uint32_t queueFamilyIndex, const Barrier &barrier);
+  void Merge(const BarrierSequence<Barrier> &other);
+  bool IsBatchEmpty(uint32_t batchIndex) const;
+  void ExtractBatch(uint32_t batchIndex, rdcarray<rdcarray<Barrier> > &result);
+  void ExtractFirstBatchForQueue(uint32_t queueFamilyIndex, rdcarray<Barrier> &result);
+  void ExtractLastBatchForQueue(uint32_t queueFamilyIndex, rdcarray<Barrier> &result);
+  inline bool empty() const { return barrierCount == 0; }
+  inline size_t size() const { return barrierCount; }
+};
+
+using ImageBarrierSequence = BarrierSequence<VkImageMemoryBarrier>;
+
+struct ImageTransitionInfo
+{
+  CaptureState capState;
+  bool separateDepthStencilLayouts;
+  inline ImageTransitionInfo(CaptureState capState, bool separateDepthStencilLayouts)
+      : capState(capState), separateDepthStencilLayouts(separateDepthStencilLayouts)
+  {
+  }
+  inline FrameRefCompFunc GetFrameRefCompFunc()
+  {
+    if(IsCaptureMode(capState))
+      return ComposeFrameRefs;
+    else
+      return KeepOldFrameRef;
+  }
+};
+
+struct ImageState
+{
+  ImageSubresourceMap subresourceStates;
+  rdcarray<VkImageMemoryBarrier> oldQueueFamilyTransfers;
+  rdcarray<VkImageMemoryBarrier> newQueueFamilyTransfers;
+  bool isMemoryBound = false;
+  ResourceId boundMemory = ResourceId();
+  VkDeviceSize boundMemoryOffset = 0ull;
+  VkDeviceSize boundMemorySize = 0ull;
+  FrameRefType maxRefType = eFrameRef_None;
+  VkImage handle = VK_NULL_HANDLE;
```

Comment:
I changed this to use a wrapped handle (and renamed it `wrappedHandle` to be extra clear).

This did require a change in `VulkanReplay::GetTextureData` that I'm not 100% sure of: there is a temporary image (`tmpImage`) that previously only had an unwrapped handle. An `ImageState` is created for `tmpImage`, to simplify the barrier creation code, and to parallel the `ImageState` operations being performed on `srcImage`. With the change to using wrapped handles, I had to create a new wrapped handle for `tmpImage` (using `GetResourceManager()->WrapResource()`).

Reply:
I can similarly change `ImageBarrierSequence` to use wrapped handles, and unwrap them immediately before recording.

Reply:
If you think having `ImageBarrierSequence` use wrapped handles wouldn't be too intrusive with wraps and unwraps then that would be preferable. If it's going to be too messy and too much code churn then I'm fine with leaving it unwrapped, since now things are named explicitly unwrapped/wrapped.